File: src/test/java/destiny/utils/LocaleComparatorTest.java
package destiny.utils;

import org.junit.Test;

import java.util.*;

/**
 * Created by smallufo on 2014-03-05.
 */
public class LocaleComparatorTest {

  @Test
  public void testCompare() throws Exception {
    Locale[] locales;
    locales = new Locale[]{Locale.ENGLISH, Locale.TAIWAN, Locale.SIMPLIFIED_CHINESE, Locale.FRANCE};

    List<Locale> list;

    System.out.println("\nDefault : " + Locale.TRADITIONAL_CHINESE);
    list = Arrays.asList(locales);
    list.sort(new LocaleComparator(Locale.TRADITIONAL_CHINESE));
    System.out.println(list);


    System.out.println("\nDefault : " + Locale.SIMPLIFIED_CHINESE);
    list = Arrays.asList(locales);
    list.sort(new LocaleComparator(Locale.SIMPLIFIED_CHINESE));
    System.out.println(list);

    System.out.println("\nDefault : " + Locale.ENGLISH);
    list = Arrays.asList(locales);
    list.sort(new LocaleComparator(Locale.ENGLISH));
    System.out.println(list);

    System.out.println("\nDefault : " + Locale.FRENCH);
    list = Arrays.asList(locales);
    list.sort(new LocaleComparator(Locale.FRENCH));
    System.out.println(list);
  }

  @Test
  public void testLangMap() throws Exception {
    Map<Locale , String> map = new HashMap<Locale , String>(){{
      put(Locale.TRADITIONAL_CHINESE, "導演");
      put(Locale.SIMPLIFIED_CHINESE,  "导演");
      put(Locale.ENGLISH, "Director");
    }};

    Locale locale;
    locale = Locale.TRADITIONAL_CHINESE;
    assert ("導演".equals(map.get(map.keySet().stream().sorted(new LocaleComparator(locale)).findFirst().get())));

    locale = Locale.SIMPLIFIED_CHINESE;
    assert ("导演".equals(map.get(map.keySet().stream().sorted(new LocaleComparator(locale)).findFirst().get())));

    locale = Locale.ENGLISH;
    assert ("Director".equals(map.get(map.keySet().stream().sorted(new LocaleComparator(locale)).findFirst().get())));

    locale = Locale.FRANCE;
    assert ("Director".equals(map.get(map.keySet().stream().sorted(new LocaleComparator(locale)).findFirst().get())));
  }
}

File: src/test/kotlin/destiny/tools/LocaleComparatorTest.kt
package destiny.tools

import java.util.*
import kotlin.test.Test
import kotlin.test.assertEquals

class LocaleComparatorTest {

  @Test
  fun testCompare() {
    val locales = arrayOf(Locale.ENGLISH, Locale.TAIWAN, Locale.SIMPLIFIED_CHINESE, Locale.FRANCE)

    println("\nDefault : " + Locale.TRADITIONAL_CHINESE)
    locales.sortedWith(LocaleComparator(Locale.TRADITIONAL_CHINESE)).also {
      assertEquals(Locale.TRADITIONAL_CHINESE , it[0])
      println(it)
    }


    println("\nDefault : " + Locale.SIMPLIFIED_CHINESE)
    locales.sortedWith(LocaleComparator(Locale.SIMPLIFIED_CHINESE)).also {
      assertEquals(Locale.SIMPLIFIED_CHINESE , it[0])
      println(it)
    }

    println("\nDefault : " + Locale.ENGLISH)
    locales.sortedWith(LocaleComparator(Locale.ENGLISH)).also {
      assertEquals(Locale.ENGLISH , it[0])
      println(it)
    }


    println("\nDefault : " + Locale.FRENCH)
    locales.sortedWith(LocaleComparator(Locale.FRENCH)).also {
      assertEquals(Locale.FRANCE , it[0])
      println(it)
    }
  }

  @Test
  fun testLangMap() {
    val map = object : HashMap<Locale, String>() {
      init {
        put(Locale.TRADITIONAL_CHINESE, "導演")
        put(Locale.SIMPLIFIED_CHINESE, "导演")
        put(Locale.ENGLISH, "Director")
      }
    }

    assert("導演" == map[map.keys.sortedWith(LocaleComparator(Locale.TRADITIONAL_CHINESE)).first()])
    assert("导演" == map[map.keys.sortedWith(LocaleComparator(Locale.SIMPLIFIED_CHINESE)).first()])
    assert("Director" == map[map.keys.sortedWith(LocaleComparator(Locale.ENGLISH)).first()])
    assert("Director" == map[map.keys.sortedWith(LocaleComparator(Locale.FRANCE)).first()])
  }
}
